ocaml/forkexecd/lib/forkhelpers.ml

```diff
@@ -60,14 +60,32 @@ let waitpid (sock, pid) =
   | _ ->
       failwith ("This should never happen: Forkhelpers.waitpid " ^ __LOC__)
 
-let waitpid_nohang ((sock, _) as x) =
+(* [waitpid_nohang] reports the status of a socket to a process. The
+   intention is to make this non-blocking. If the process is finished,
+   the socket is closed by default and not otherwise. The [close] option
+   can be used to never close the socket such that this is left to the
+   caller, which is a cleaner protocol. *)
+let waitpid_nohang ?(close = true) (sock, pid) =
+  let fail fmt = Printf.kprintf failwith fmt in
   Unix.set_nonblock sock ;
-  let r =
-    try waitpid x
-    with Unix.(Unix_error ((EAGAIN | EWOULDBLOCK), _, _)) ->
-      (0, Unix.WEXITED 0)
-  in
-  Unix.clear_nonblock sock ; r
+  match Fecomms.read_raw_rpc sock with
+  | Fe.Finished (Fe.WEXITED n) ->
+      if close then Unix.close sock ;
+      (pid, Unix.WEXITED n)
+  | Fe.Finished (Fe.WSIGNALED n) ->
+      if close then Unix.close sock ;
+      (pid, Unix.WSIGNALED n)
+  | Fe.Finished (Fe.WSTOPPED n) ->
+      if close then Unix.close sock ;
+      (pid, Unix.WSTOPPED n)
+  | exception Unix.(Unix_error ((EAGAIN | EWOULDBLOCK), _, _)) ->
+      Unix.clear_nonblock sock ;
```

Comment:
This is a superset of https://github.com/xapi-project/xen-api/pull/5067, so closing that one, but see that PR for more discussions on what is wrong with forkexec usage.

Reply:
To summarize that PR, I think I was worried there that the approach with closing the socket or not is too difficult to get right (so we may introduce an EBADF elsewhere by fixing this one), and I think my latest thinking was that we should change the protocol to remove the need for this weird protocol where sometimes the socket needs to stay open or not, but I haven't had time to work on that idea yet:

> I'd like to completely remove the open socket from here, and e.g. establish a fresh one to query forkexecd about the status of the process instead of keeping one open, but that has other drawbacks (it prevents FD leaks in the client, but may leave an FD/memory leak in forkexecd if we never 'wait' for a process that exited), I'll need to think about this more to see whether I find an alternative.